src/timeoffset.rs

```diff
@@ -0,0 +1,362 @@
+use crate::{Epoch, TimeScale, Unit};
+
+/// Errors during [TimeOffset] exploitation
+#[derive(Debug)]
+pub enum TimeOffsetError {
+    /// Timescales should not be identical when defining a new [TimeOffset]!
+    IdenticalTimescales,
+    /// Epoch is not expressed in one of the two supported [TimeScale]s!
+    NotSupportedTimescale,
+    /// Epoch should be expressed in the left-hand side [TimeScale].
+    InvalidTimescale,
+    /// [TimeOffset] is outdated and should not apply (difference is too large): should be updated weekly at least!
+    OutdatedTimeOffset,
+}
+
+/// [TimeOffset] used in [TimeShift]ing operations
+#[derive(Copy, Clone, PartialEq)]
+pub struct TimeOffset {
+    /// Left hand side (compared) [TimeScale]
+    lhs: TimeScale,
+    /// Right hand side (reference) [TimeScale]
+    rhs: TimeScale,
+    /// Weekly reference time (counter, nanoseconds)
+    t_ref_nanos: (u32, u64),
+    /// Polynomials (s, s.s⁻¹, s.s⁻²)
+    polynomials: (f64, f64, f64),
+}
+
+impl TimeOffset {
+    /// Define a new [TimeOffset] from this reference [Epoch] expressed
+    /// in left-hand side [TimeScale] to rhs [TimeScale]
+    pub fn from_reference_epoch(
+        t_ref: Epoch,
+        rhs: TimeScale,
+        polynomials: (f64, f64, f64),
+    ) -> Result<Self, TimeOffsetError> {
+        if t_ref.time_scale == rhs {
+            return Err(TimeOffsetError::IdenticalTimescales);
+        }
+
+        let t_ref_nanos = t_ref.to_time_of_week();
+
+        Ok(Self {
+            rhs,
+            t_ref_nanos,
+            polynomials,
+            lhs: t_ref.time_scale,
+        })
+    }
+
+    /// Define a new [TimeOffset] from reference time of week and other components
+    pub fn from_reference_time_of_week(
+        t_ref_nanos: (u32, u64),
+        lhs: TimeScale,
+        rhs: TimeScale,
+        polynomials: (f64, f64, f64),
+    ) -> Result<Self, TimeOffsetError> {
+        if lhs == rhs {
+            return Err(TimeOffsetError::IdenticalTimescales);
+        }
+
+        Ok(Self {
+            lhs,
+            rhs,
+            t_ref_nanos,
+            polynomials,
+        })
+    }
+
+    /// Define a new |[TimeScale::GPST] - [TimeScale::UTC]| [TimeOffset] from time of week components
+    pub fn from_gpst_utc_time_of_week(
```

Comment:
In the original API, we had a lot of initializers for all of the different time scales: let's avoid this mistake again, it was difficult to remove once added. Let's expect the users to use the generic initializer and correctly build their own epoch before calling it.